File: web/src/api/operator.ts
import apiClient from './client'

export type OperatorCategory = 'analysis' | 'processing' | 'generation' | 'utility'
export type OperatorStatus = 'draft' | 'testing' | 'published' | 'deprecated'
export type OperatorOrigin = 'builtin' | 'custom' | 'marketplace' | 'mcp'
export type OperatorExecMode = 'http' | 'cli' | 'mcp' | 'ai_model'

export interface OperatorVersion {
  id: string
  version: string
  exec_mode: OperatorExecMode
  exec_config?: Record<string, any>
  status: 'draft' | 'active' | 'archived'
}

export interface Operator {
  id: string
  code: string
  name: string
  description?: string
  category: OperatorCategory
  type: string
  origin?: OperatorOrigin
  active_version_id?: string
  exec_mode?: OperatorExecMode
  active_version?: OperatorVersion
  version: string
  endpoint: string
  method: string
  input_schema?: Record<string, any>
  output_spec?: Record<string, any>
  config?: Record<string, any>
  status: OperatorStatus
  is_builtin: boolean
  tags?: string[]
  visibility?: number
  visible_role_ids?: string[]
  created_at: string
  updated_at: string
}

export interface OperatorListQuery {
  category?: OperatorCategory
  type?: string
  status?: OperatorStatus
  origin?: OperatorOrigin
  exec_mode?: OperatorExecMode
  is_builtin?: boolean
  tags?: string[]
  keyword?: string
  page?: number
  page_size?: number
}

export interface OperatorCreateReq {
  code: string
  name: string
  description?: string
  category: OperatorCategory
  type: string
  origin?: OperatorOrigin
  exec_mode?: OperatorExecMode
  exec_config?: Record<string, any>
  version?: string
  endpoint?: string
  method?: string
  input_schema?: Record<string, any>
  output_spec?: Record<string, any>
  config?: Record<string, any>
  status?: OperatorStatus
  is_builtin?: boolean
  tags?: string[]
  visibility?: number
  visible_role_ids?: string[]
}

export interface OperatorUpdateReq {
  name?: string
  description?: string
  category?: OperatorCategory
  tags?: string[]
  visibility?: number
  visible_role_ids?: string[]
}

export interface OperatorListResponse {
  items: Operator[]
  total: number
}

export interface OperatorVersionListQuery {
  page?: number
  page_size?: number
}

export interface OperatorVersionListResponse {
  items: OperatorVersion[]
  total: number
}

export interface TestOperatorReq {
  asset_id?: string
  params?: Record<string, any>
}

export interface TestOperatorResponse {
  success: boolean
  message: string
  diagnostics?: Record<string, any>
}

export interface OperatorVersionCreateReq {
  version: string
  exec_mode: OperatorExecMode
  exec_config?: Record<string, any>
  input_schema?: Record<string, any>
  output_spec?: Record<string, any>
  config?: Record<string, any>
  changelog?: string
  status?: string
}

export interface OperatorVersionActionReq {
  version_id: string
}

export interface ValidateSchemaReq {
  schema: Record<string, any>
}

export interface ValidateConnectionReq {
  upstream_output_spec: Record<string, any>
  downstream_input_schema: Record<string, any>
}

export interface ValidateResultResponse {
  valid: boolean
  message?: string
}

export interface OperatorTemplate {
  id: string
  code: string
  name: string
  description?: string
  category: OperatorCategory
  type: string
  exec_mode: OperatorExecMode
  exec_config?: Record<string, any>
  input_schema?: Record<string, any>
  output_spec?: Record<string, any>
  config?: Record<string, any>
  author?: string
  tags?: string[]
  icon_url?: string
  downloads: number
  created_at: string
  updated_at: string
}

export interface TemplateListQuery {
  category?: OperatorCategory
  type?: string
  exec_mode?: OperatorExecMode
  tags?: string[]
  keyword?: string
  page?: number
  page_size?: number
}

export interface OperatorTemplateListResponse {
  items: OperatorTemplate[]
  total: number
}

export interface InstallTemplateReq {
  template_id: string
  operator_code: string
  operator_name: string
  tags?: string[]
}

export interface OperatorDependency {
  id: string
  operator_id: string
  depends_on_id: string
  min_version?: string
  is_optional: boolean
  created_at: string
}

export interface SetDependenciesReq {
  dependencies: Array<{
    depends_on_id: string
    min_version?: string
    is_optional?: boolean
  }>
}

export interface DependencyCheckResponse {
  satisfied: boolean
  unmet?: string[]
}

export interface MCPServer {
  id: string
  name: string
  description?: string
  status?: string
}

export interface MCPTool {
  name: string
  description?: string
  version?: string
  input_schema?: Record<string, any>
  output_schema?: Record<string, any>
}

export interface MCPInstallReq {
  server_id: string
  tool_name: string
  operator_code: string
  operator_name: string
  category?: OperatorCategory
  type?: string
  timeout_sec?: number
  tags?: string[]
}

export interface SyncMCPTemplatesReq {
  server_id: string
}

export interface SyncMCPTemplatesResponse {
  server_id: string
  total: number
  created: number
  updated: number
}

function toListParams(params?: { page?: number; page_size?: number; [key: string]: any }) {
  if (!params) return undefined
  const { page, page_size, ...rest } = params
  const normalized: Record<string, any> = { ...rest }
  if (typeof page_size === 'number' && page_size > 0) {
    normalized.limit = page_size
  }
  if (typeof page === 'number' && page > 0 && typeof page_size === 'number' && page_size > 0) {
    normalized.offset = (page - 1) * page_size
  }
  return normalized
}

export const operatorApi = {
  list(params?: OperatorListQuery) {
    return apiClient.get<OperatorListResponse>('/operators', { params: toListParams(params) })
  },

  get(id: string) {
    return apiClient.get<Operator>(`/operators/${id}`)
  },

  create(data: OperatorCreateReq) {
    return apiClient.post<Operator>('/operators', data)
  },

  update(id: string, data: OperatorUpdateReq) {
    return apiClient.put<Operator>(`/operators/${id}`, data)
  },

  delete(id: string) {
    return apiClient.delete(`/operators/${id}`)
  },

  publish(id: string) {
    return apiClient.post<Operator>(`/operators/${id}/publish`)
  },

  deprecate(id: string) {
    return apiClient.post<Operator>(`/operators/${id}/deprecate`)
  },

  test(id: string, data?: TestOperatorReq) {
    return apiClient.post<TestOperatorResponse>(`/operators/${id}/test`, data ?? {})
  },

  listVersions(id: string, params?: OperatorVersionListQuery) {
    return apiClient.get<OperatorVersionListResponse>(`/operators/${id}/versions`, { params: toListParams(params) })
  },

  getVersion(id: string, versionId: string) {
    return apiClient.get<OperatorVersion>(`/operators/${id}/versions/${versionId}`)
  },

  createVersion(id: string, data: OperatorVersionCreateReq) {
    return apiClient.post<OperatorVersion>(`/operators/${id}/versions`, data)
  },

  activateVersion(id: string, data: OperatorVersionActionReq) {
    return apiClient.post<Operator>(`/operators/${id}/versions/activate`, data)
  },

  rollbackVersion(id: string, data: OperatorVersionActionReq) {
    return apiClient.post<Operator>(`/operators/${id}/versions/rollback`, data)
  },

  archiveVersion(id: string, data: OperatorVersionActionReq) {
    return apiClient.post<OperatorVersion>(`/operators/${id}/versions/archive`, data)
  },

  validateSchema(data: ValidateSchemaReq) {
    return apiClient.post<ValidateResultResponse>('/operators/validate-schema', data)
  },

  validateConnection(data: ValidateConnectionReq) {
    return apiClient.post<ValidateResultResponse>('/operators/validate-connection', data)
  },

  listTemplates(params?: TemplateListQuery) {
    return apiClient.get<OperatorTemplateListResponse>('/operators/templates', { params: toListParams(params) })
  },

  getTemplate(templateId: string) {
    return apiClient.get<OperatorTemplate>(`/operators/templates/${templateId}`)
  },

  installTemplate(data: InstallTemplateReq) {
    return apiClient.post<Operator>('/operators/templates/install', data)
  },

  listDependencies(id: string) {
    return apiClient.get<OperatorDependency[]>(`/operators/${id}/dependencies`)
  },

  setDependencies(id: string, data: SetDependenciesReq) {
    return apiClient.put(`/operators/${id}/dependencies`, data)
  },

  checkDependencies(id: string) {
    return apiClient.get<DependencyCheckResponse>(`/operators/${id}/dependencies/check`)
  },

  listMCPServers() {
    return apiClient.get<MCPServer[]>('/operators/mcp/servers')
  },

  listMCPTools(serverId: string) {
    return apiClient.get<MCPTool[]>(`/operators/mcp/servers/${serverId}/tools`)
  },

  previewMCPTool(serverId: string, toolName: string) {
    return apiClient.get<MCPTool>(`/operators/mcp/servers/${serverId}/tools/${encodeURIComponent(toolName)}/preview`)
  },

  installMCPOperator(data: MCPInstallReq) {
    return apiClient.post<Operator>('/operators/mcp/install', data)
  },

  syncMCPTemplates(data: SyncMCPTemplatesReq) {
    return apiClient.post<SyncMCPTemplatesResponse>('/operators/mcp/sync-templates', data)
  },

  listByCategory(category: string, params?: OperatorListQuery) {
    return apiClient.get<Operator[]>(`/operators/category/${category}`, { params: toListParams(params) })
  }
}

File: internal/api/dto/operator.go
package dto

import (
	"time"

	"goyavision/internal/domain/operator"

	"github.com/google/uuid"
)

// OperatorListQuery 列出算子查询参数
type OperatorListQuery struct {
	Category  *string `query:"category"`
	Type      *string `query:"type"`
	Status    *string `query:"status"`
	Origin    *string `query:"origin"`
	ExecMode  *string `query:"exec_mode"`
	IsBuiltin *bool   `query:"is_builtin"`
	Tags      *string `query:"tags"`
	Keyword   *string `query:"keyword"`
	Limit     int     `query:"limit"`
	Offset    int     `query:"offset"`
}

// OperatorCreateReq 创建算子请求
type OperatorCreateReq struct {
	Code        string                 `json:"code" validate:"required"`
	Name        string                 `json:"name" validate:"required"`
	Description string                 `json:"description,omitempty"`
	Category    string                 `json:"category" validate:"required"`
	Type        string                 `json:"type" validate:"required"`
	Origin      string                 `json:"origin,omitempty"`
	ExecMode    string                 `json:"exec_mode,omitempty"`
	ExecConfig  map[string]interface{} `json:"exec_config,omitempty"`
	Version     string                 `json:"version,omitempty"`
	Endpoint    string                 `json:"endpoint,omitempty"`
	Method      string                 `json:"method,omitempty"`
	InputSchema map[string]interface{} `json:"input_schema,omitempty"`
	OutputSpec  map[string]interface{} `json:"output_spec,omitempty"`
	Config      map[string]interface{} `json:"config,omitempty"`
	Status      string                 `json:"status,omitempty"`
	IsBuiltin   bool                   `json:"is_builtin,omitempty"`
	Tags        []string               `json:"tags,omitempty"`
}

// OperatorUpdateReq 更新算子请求
type OperatorUpdateReq struct {
	Name        *string                `json:"name,omitempty"`
	Description *string                `json:"description,omitempty"`
	Category    *string                `json:"category,omitempty"`
	Tags        []string               `json:"tags,omitempty"`
}

type TestOperatorReq struct {
	AssetID *uuid.UUID             `json:"asset_id,omitempty"`
	Params  map[string]interface{} `json:"params,omitempty"`
}

type TestOperatorResponse struct {
	Success     bool                   `json:"success"`
	Message     string                 `json:"message"`
	Diagnostics map[string]interface{} `json:"diagnostics,omitempty"`
}

type MCPServerResponse struct {
	ID          string `json:"id"`
	Name        string `json:"name"`
	Description string `json:"description,omitempty"`
	Status      string `json:"status,omitempty"`
}

type MCPToolResponse struct {
	Name         string                 `json:"name"`
	Description  string                 `json:"description,omitempty"`
	Version      string                 `json:"version,omitempty"`
	InputSchema  map[string]interface{} `json:"input_schema,omitempty"`
	OutputSchema map[string]interface{} `json:"output_schema,omitempty"`
}

type MCPInstallReq struct {
	ServerID     string   `json:"server_id" validate:"required"`
	ToolName     string   `json:"tool_name" validate:"required"`
	OperatorCode string   `json:"operator_code" validate:"required"`
	OperatorName string   `json:"operator_name" validate:"required"`
	Category     *string  `json:"category,omitempty"`
	Type         *string  `json:"type,omitempty"`
	TimeoutSec   int      `json:"timeout_sec,omitempty"`
	Tags         []string `json:"tags,omitempty"`
}

type SyncMCPTemplatesReq struct {
	ServerID string `json:"server_id" validate:"required"`
}

type SyncMCPTemplatesResponse struct {
	ServerID string `json:"server_id"`
	Total    int    `json:"total"`
	Created  int    `json:"created"`
	Updated  int    `json:"updated"`
}

type OperatorVersionResponse struct {
	ID         uuid.UUID              `json:"id"`
	Version    string                 `json:"version"`
	ExecMode   string                 `json:"exec_mode"`
	ExecConfig map[string]interface{} `json:"exec_config,omitempty"`
	Status     string                 `json:"status"`
}

// OperatorResponse 算子响应
type OperatorResponse struct {
	ID          uuid.UUID              `json:"id"`
	Code        string                 `json:"code"`
	Name        string                 `json:"name"`
	Description string                 `json:"description,omitempty"`
	Category    string                 `json:"category"`
	Type        string                 `json:"type"`
	Origin      string                 `json:"origin,omitempty"`
	ActiveVersionID *uuid.UUID         `json:"active_version_id,omitempty"`
	ExecMode    string                 `json:"exec_mode,omitempty"`
	ActiveVersion *OperatorVersionResponse `json:"active_version,omitempty"`
	Version     string                 `json:"version"`
	Endpoint    string                 `json:"endpoint"`
	Method      string                 `json:"method"`
	InputSchema map[string]interface{} `json:"input_schema,omitempty"`
	OutputSpec  map[string]interface{} `json:"output_spec,omitempty"`
	Config      map[string]interface{} `json:"config,omitempty"`
	Status      string                 `json:"status"`
	IsBuiltin   bool                   `json:"is_builtin"`
	Tags        []string               `json:"tags,omitempty"`
	CreatedAt   time.Time              `json:"created_at"`
	UpdatedAt   time.Time              `json:"updated_at"`
}

// OperatorListResponse 算子列表响应
type OperatorListResponse struct {
	Items []*OperatorResponse `json:"items"`
	Total int64               `json:"total"`
}

// OperatorToResponse 转换为响应
func OperatorToResponse(o *operator.Operator) *OperatorResponse {
	if o == nil {
		return nil
	}

	inputSchema := o.InputSchema
	if inputSchema == nil {
		inputSchema = make(map[string]interface{})
	}

	outputSpec := o.OutputSpec
	if outputSpec == nil {
		outputSpec = make(map[string]interface{})
	}

	config := o.Config
	if config == nil {
		config = make(map[string]interface{})
	}

	tags := o.Tags
	if tags == nil {
		tags = []string{}
	}

	var activeVersion *OperatorVersionResponse
	execMode := ""
	if o.ActiveVersion != nil {
		execMode = string(o.ActiveVersion.ExecMode)
		activeVersion = &OperatorVersionResponse{
			ID:      o.ActiveVersion.ID,
			Version: o.ActiveVersion.Version,
			ExecMode: string(o.ActiveVersion.ExecMode),
			Status:  string(o.ActiveVersion.Status),
		}
	}

	return &OperatorResponse{
		ID:          o.ID,
		Code:        o.Code,
		Name:        o.Name,
		Description: o.Description,
		Category:    string(o.Category),
		Type:        string(o.Type),
		Origin:      string(o.Origin),
		ActiveVersionID: o.ActiveVersionID,
		ExecMode:    execMode,
		ActiveVersion: activeVersion,
		Version:     o.Version,
		Endpoint:    o.Endpoint,
		Method:      o.Method,
		InputSchema: inputSchema,
		OutputSpec:  outputSpec,
		Config:      config,
		Status:      string(o.Status),
		IsBuiltin:   o.IsBuiltin,
		Tags:        tags,
		CreatedAt:   o.CreatedAt,
		UpdatedAt:   o.UpdatedAt,
	}
}

// OperatorsToResponse 转换为响应列表
func OperatorsToResponse(operators []*operator.Operator) []*OperatorResponse {
	result := make([]*OperatorResponse, len(operators))
	for i, o := range operators {
		result[i] = OperatorToResponse(o)
	}
	return result
}
